File: backend/app/schemas/finance.py
from datetime import datetime
from decimal import Decimal
from typing import Optional

from pydantic import BaseModel, Field

from ..models import ExpenseCategory


class ExpenseCreateRequest(BaseModel):
    category: ExpenseCategory
    amount: Decimal = Field(gt=0, max_digits=12, decimal_places=2)
    description: Optional[str] = None
    branch_id: Optional[int] = None


class ExpenseResponse(BaseModel):
    id: int
    category: ExpenseCategory
    amount: Decimal
    description: Optional[str]
    branch_id: Optional[int]
    created_by_name: str
    created_at: datetime


class FinancialsResponse(BaseModel):
    # CLAUDE.md §30: derived read-only from Transaction (income) +
    # ExpenseEntry (expenses) — never a second ledger. Transaction doesn't
    # split a repayment into its principal/interest components (that split
    # only exists per-installment on RepaymentSchedule), so income here is
    # reported the two ways Transaction actually itemizes it: repayments
    # collected and penalties collected.
    period_start: str
    total_repayment_income: Decimal
    total_penalty_income: Decimal
    total_disbursed: Decimal
    total_expenses: Decimal
    net: Decimal
    recent_expenses: list[ExpenseResponse]

File: frontend/src/schemas/finance.ts
import { z } from "zod";

// Mirrors backend app/models/expense_entry.py::ExpenseCategory
export const EXPENSE_CATEGORIES = [
  "salaries",
  "rent",
  "transport",
  "communications",
  "administration",
  "other",
] as const;

export const EXPENSE_CATEGORY_LABELS: Record<string, string> = {
  salaries: "Salaries",
  rent: "Rent",
  transport: "Transport",
  communications: "Communications",
  administration: "Administration",
  other: "Other",
};

// Mirrors backend app/schemas/finance.py
export const expenseCreateSchema = z.object({
  category: z.enum(EXPENSE_CATEGORIES),
  amount: z.string().min(1, "Amount is required"),
  description: z.string().optional(),
});
export type ExpenseCreateInput = z.infer<typeof expenseCreateSchema>;

export interface ExpenseResponse {
  id: number;
  category: string;
  amount: string;
  description: string | null;
  branch_id: number | null;
  created_by_name: string;
  created_at: string;
}

export interface FinancialsResponse {
  period_start: string;
  total_repayment_income: string;
  total_penalty_income: string;
  total_disbursed: string;
  total_expenses: string;
  net: string;
  recent_expenses: ExpenseResponse[];
}
